vignettes/new-guides/transfer_learning.Rmd

```diff
@@ -0,0 +1,612 @@
+only process contiguous batches of data), and we'll do the input value scaling as part
+ of the model, when we create it.
+
+Let's resize images to 150x150:
+
+```{r}
+library(magrittr, include.only = "%<>%")
+size <- as.integer(c(150, 150))
+train_ds      %<>% dataset_map(function(x, y) list(tf$image$resize(x, size), y))
+validation_ds %<>% dataset_map(function(x, y) list(tf$image$resize(x, size), y))
+test_ds       %<>% dataset_map(function(x, y) list(tf$image$resize(x, size), y))
+```
+
+Besides, let's batch the data and use caching & prefetching to optimize loading speed.
+```{r}
+dataset_cache_batch_prefetch <- function(dataset, batch_size = 32, buffer_size = 10) {
+  dataset %>% 
+    dataset_cache() %>% 
+    dataset_batch(batch_size) %>% 
+    dataset_prefetch(buffer_size)
+}
+
+train_ds      %<>% dataset_cache_batch_prefetch()
+validation_ds %<>% dataset_cache_batch_prefetch()
+test_ds       %<>% dataset_cache_batch_prefetch()
+```
+
+### Using random data augmentation
+
+When you don't have a large image dataset, it's a good practice to artificially
+ introduce sample diversity by applying random yet realistic transformations to
+the training images, such as random horizontal flipping or small random rotations. This
+helps expose the model to different aspects of the training data while slowing down
+ overfitting.
+ 
+```{r}
+data_augmentation <- keras_model_sequential() %>% 
+  layer_random_flip("horizontal") %>% 
+  layer_random_rotation(.1)
+```
+
+Let's visualize what the first image of the first batch looks like after various random
+ transformations:
+
+```{r}
+batch <- train_ds %>% 
+  dataset_take(1) %>% 
+  as_iterator() %>% iter_next()
+
+c(images, labels) %<-% batch
+first_image <- images[1, all_dims(), drop = FALSE]
+augmented_image <- data_augmentation(first_image, training = TRUE)
+
+plot_image <- function(image, main = deparse1(substitute(image))) {
+  image %>%
+    k_squeeze(1) %>% # drop batch dim
+    as.array() %>%   # convert from tensor to R array
+    as.raster(max = 255) %>%
+    plot()
+  
+  if(!is.null(main)) 
+    title(main)
+}
+
+par(mfrow = c(2, 2), mar = c(1, 1, 1.5, 1))
+plot_image(first_image)
+plot_image(augmented_image)
+plot_image(data_augmentation(first_image, training = TRUE), "augmented 2")
+plot_image(data_augmentation(first_image, training = TRUE), "augmented 3")
+# 
+# augmented_image %>% 
+#   k_squeeze(1) %>% # drop batch dim
+#   as.array() %>% as.raster(max = 255) %>% 
+#   plot()
+# title(as.array(labels[1]))
+
+```
+
+
+
+
+## Build a model
+
+Now let's built a model that follows the blueprint we've explained earlier.
+
+Note that:
+
+- We add `layer_rescaling()` to scale input values (initially in the `[0, 255]`
+ range) to the `[-1, 1]` range.
+- We add a `layer_dropout()` before the classification layer, for regularization.
+- We make sure to pass `training = FALSE` when calling the base model, so that
```

Comment:
hm - now I think my construction above was wrong, and it really should have been `training=FALSE` there too  (what I first thought)?
Indeed that would be way less confusing ...